homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,306 @@
+"""Support for OSO Energy water heaters."""
+from collections.abc import Mapping
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.water_heater import (
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from . import OSOEnergyEntity, refresh_system
+from .const import (
+    ATTR_FULL_UTILIZATION,
+    ATTR_PROFILE_HOURS,
+    ATTR_V40MIN,
+    DOMAIN,
+    EXTRA_HEATER_ATTR,
+    HEATER_MAX_TEMP,
+    HEATER_MIN_TEMP,
+    MANUFACTURER,
+    OPERATION_LIST,
+    OSO_ENERGY_TO_HASS_STATE,
+    SERVICE_SET_PROFILE,
+    SERVICE_SET_V40MIN,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up OSO Energy heater based on a config entry."""
+    osoenergy = hass.data[DOMAIN][entry.entry_id]
+    devices = osoenergy.session.device_list.get("water_heater")
+    entities = []
+    if devices:
+        for dev in devices:
+            entities.append(OSOEnergyWaterHeater(osoenergy, dev))
+    async_add_entities(entities, True)
+
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_TURN_ON,
+        {vol.Required(ATTR_FULL_UTILIZATION): vol.All(cv.boolean)},
+        "async_oso_turn_on",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_TURN_OFF,
+        {vol.Required(ATTR_FULL_UTILIZATION): vol.All(cv.boolean)},
```

Comment:
In the case of the OSO Energy heaters the devices themselves cannot be turned off in the sense of switching it off like other devices. 
There is no off/idle state for the heaters themselves - they remain on and connected all the time. 
The on/off state indicates if the heater is heating the water at the moment. 
The users can turn it off - stop the device from heating the water for one hour if the full utilization parameter is false and until a minimum level of energy is reached when the full utilization is true. 
It's the same case for switching it on - force the heater to heat the water for one hour or until the maximum energy is reached. 

For the water heater services turn_on and turn_off that come from Home Asistant these options are implemented by setting the full utilization to true.

Reply:
I do not fully understand how ATTR_FULL_UTILIZATION works. Do you think an owner of this device will understand?

Reply:
I've added an explanation to the full utilization in the documentation of the integration.
The best explanation I can give is that turning a heater on is the equivalent to the boost mode in some heaters, while turn off is the equivalent of power save or the opposite of the boost mode. The full utilization specifies the amount of time the heater stays in this mode - false for one hour or true if the user decides that the heater must operate in the selected mode until the minimum or maximum temperature is reached.
This would provide a way for the users to preheat the water in the heater if they expect a more intensive usage of hot water or disable the heating for a period of time - one hour or until minimum temperature is reached - for example to reduce the power consumption at expensive hours.

Reply:
```
turn_on:
  name: Turn on Heating
  description: Turn on Heating for one hour or until max temperature is reached
  target:
    entity:
      domain: water_heater
  fields:
    full_utilization:
      name: Full Utilization
      description: Choose if heating should be on until max temperature (True) is reached or for one hour (False)
      required: true
      default: false
      example: false
      selector:
        boolean:
turn_off:
  name: Turn off Heating
  description: Turn off Heating for one hour or until min temperature is reached
  target:
    entity:
      domain: water_heater
  fields:
    full_utilization:
      name: Full Utilization
      description: Choose if heating should be off until min temperature (True) is reached or for one hour (False)
```

Could the "full_utilization" setting be named something like "until_temp_limit" or similar to better explain its usage?

At least that's how I understand the usage of this setting. E.g. the on or off state is normally for one hour, but with "full utilization" it'll be on or off _until_ it reaches the temperature limit (= max for on, min for off)?

Reply:
As @oddeirik already has commented, I also recommend renaming it differently. `until_temp_limit` makes a lot more sense to me.